Add documentation for "Running Containers" (#19982)
docker/docs: content/guides/docker-concepts/running-containers/overriding-container-defaults.md

### Restricting the container to consume the resources

You can use the `--memory` and `--cpus` flags with the `docker run` command to restrict how much CPU and memory a container can use. For example, you can set a memory limit for the Python API container, preventing it from consuming excessive resources on your host. Here's the command:

```console
$ docker run --memory="512m" --cpus="0.5" my_image
 ```

This command limits container memory usage to 512 megabytes and defines the CPU quota of 0.5 for half a core.

> **Monitor the real-time resource usage**
>
> You can use the `docker stats` command to monitor the real-time resource usage of running containers. This helps you understand whether the allocated resources are sufficient or need adjustment.
{ .tip }

By effectively using these `docker run` flags, you can tailor your containerized application's behaviour to fit your specific requirements!

## Try it now

In this hands-on guide, you'll see how to use the `docker run` command to override the container defaults.

> 1. [Download and install](https://www.docker.com/products/docker-desktop/) Docker Desktop.

Reviewer: Either remove the numbered list here, or add this as the first step in the section below.
```suggestion
1. [Download and install](/get-docker/) Docker Desktop.
```

### Run the multiple instance of Redis database

1.  Start a container using the [Redis image](https://hub.docker.com/_/redis) with the following command:
    
    ```console
    $ docker run -d -p 6379:6379 redis
    ```

    This will start the Redis database in the background, listening on the standard container port `6379` and mapped to port `6379` on the host machine.

2. Start a new Redis container mapped to the different port

    ```console
    $ docker run -d -p 6380:6379 redis
    ```

    This will start a new Redis container in the background, listening on the standard container port `6379` but mapped to port `6380` on the host machine. You override the host port just to ensure that this new container doesn't conflict with the existing running container.

> 3. Verify if both containers are running via the Docker dashboard

Reviewer: ```suggestion
3. Verify if both containers are running via the Docker Dashboard.
```

Here's where the `--network` flag with the `docker run` command comes in.

Follow the steps to see how to connect a Redis container to a custom network.

1. Create a network by using the following command:

    ```console
    $ docker network create mynetwork
    ```

2. Verify the network by running the following command:

    ```console
    $ docker network ls
    ```

3. Connect Redis to the existing network by using the following command:

    ```console
    $ docker run -d -p 6381:6379 –network mynetwork redis
    ```

    This will start Redis container in the background, mapped to the host port 6381 and attached to the `mynetwork` network. You passed the `--network` parameter to override the container default by connecting the container to custom Docker network for better isolation and communication with other containers.

    > **Inspecting the container network**
    >
    > You can use `docker network inspect` command to see if the container is tied to this new bridge network.
    { .tip }

### Manage the resources

> By default, containers are not limited in their resource usage. However, on shared systems, it's crucial to manage resources effectively. It's important not to allow a running container to consume too much of the host machine's memory.

Reviewer: ```suggestion
By default, containers are not limited in their resource usage. However, on shared systems, it's crucial to manage resources effectively. It's important not to let a running container consume too much of the host machine's memory.
```

This is where the `docker run` command shines again. It offers flags like `--memory` and `--cpus` to restrict how much CPU and memory a container can use.

```console
$ docker run -d --memory=”512m” --cpus=”.5” redis
```

> The `--cpus` flag specifies the CPU quota for the container. Here, it's set to half a CPU core (0.5) whereas the `--memory` flag specifies the memory limit for the container. In this case, it's set to 512 megabytes (512MB).

Reviewer: ```suggestion
The `--cpus` flag specifies the CPU quota for the container. Here, it's set to half a CPU core (0.5) whereas the `--memory` flag specifies the memory limit for the container. In this case, it's set to 512 MB.
```

### Override the default `CMD` and `ENTRYPOINT` in Docker Compose

> Sometimes, you might need to override the default commands (CMD) or entry points (ENTRYPOINT) defined in a Docker image, especially when using Docker Compose.

Reviewer: ```suggestion
Sometimes, you might need to override the default commands (`CMD`) or entry points (`ENTRYPOINT`) defined in a Docker image, especially when using Docker Compose.
```

> 1. Create a `docker-compose.yaml` file with the following content:

Reviewer: ```suggestion
1. Create a `compose.yaml` file with the following content:
```

    ```yaml
    services:
        redis:
         image: redis
         entrypoint: ["docker-entrypoint.sh"]
         command: ["redis-server", "--requirepass", "redispassword"]
    ```

>     The YAML file defines a service named `redis` that uses the official Redis image, sets an entrypoint script, and starts the container with password authentication.

Reviewer: ```suggestion
    The Compose file defines a service named `redis` that uses the official Redis image, sets an entrypoint script, and starts the container with password authentication.
```

2. Bring up the service by running the following command:

    ```console
    $ docker compose up -d
    ```

    This command starts the Redis service defined in the Docker Compose file.

> 3. Verify the authentication with Docker Dashboard

Reviewer: ```suggestion
3. Verify the authentication with Docker Dashboard.
```

>     Open the **Docker Dashboard**, select the **Redis** container and select **Exec** to enter into the container shell.

Reviewer: ```suggestion
    Open the Docker Dashboard, select the **Redis** container and select **Exec** to enter into the container shell.
```

    ![A screenshot of the Docker Dashboard selecting the Redis container and entering into its shell using EXEC button](images/exec-into-redis-container.webp?border=true) 

## Additional resources

> * [Build Variables](https://docs.docker.com/build/building/variables/)

Reviewer: ```suggestion
* [Build variables](/build/building/variables/)
```

> * [Ways to set environment variables with Compose](https://docs.docker.com/compose/environment-variables/set-environment-variables/)

Reviewer: ```suggestion
* [Ways to set environment variables with Compose](/compose/environment-variables/set-environment-variables/)
```

Now you might ask, "How can I override these container defaults?" The answer is `docker run` command.

The `docker run` command offers a powerful way to override these defaults and tailor the container's behavior to your liking. The command offers several flags that let you to customize container behavior on the fly.

Here's a few ways you can achieve this.

### Overriding the network ports

Sometimes you might want to use separate database instances for development and testing purposes. Running these database instances on the same port might conflict. You can use the `-p` option in `docker run` to map container ports to host ports, allowing you to run the multiple instances of the container without any conflict.

```console

> $ docker run -d -p HOST_PORT:CONTAINER_PORT my_image

Reviewer 2: I don't think we should use `my_image` here and in the following examples. Let's use `alpine` instead (and add a brief note about it in the introduction)

Author: As the initial section talks about the database, I picked up postgres Docker image instead rather than my_image. Do check if it looks good. @dvdksn

Author: @dvdksn I have changed all the my_image to the real world Docker images. Can you please have a look?

{ .tip }

### Restricting the container to consume the resources

You can use the `--memory` and `--cpus` flags with the `docker run` command to restrict how much CPU and memory a container can use. For example, you can set a memory limit for the Python API container, preventing it from consuming excessive resources on your host. Here's the command:

```console
$ docker run --memory="512m" --cpus="0.5" my_image
 ```

This command limits container memory usage to 512 MB and defines the CPU quota of 0.5 for half a core.

> **Monitor the real-time resource usage**
>
> You can use the `docker stats` command to monitor the real-time resource usage of running containers. This helps you understand whether the allocated resources are sufficient or need adjustment.
{ .tip }

By effectively using these `docker run` flags, you can tailor your containerized application's behavior to fit your specific requirements.

## Try it now

In this hands-on guide, you'll see how to use the `docker run` command to override the container defaults.

1. [Download and install](/get-docker/) Docker Desktop.

### Run the multiple instance of Redis database

> 1.  Start a container using the [Redis image](https://hub.docker.com/_/redis) with the following command:

Reviewer 2: Given the recent licensing events about redis, let's maybe use a different example image. I propose `memcached`

Author: @dvdksn I need to test it with Memcached. Will test it tomorrow and let you know.

Author: I have pushed the revised guide replacing redis with postgres.

Author: Also, I am checking internally to see if we can replace the video with Postgres example.

3. Verify if both containers are running via the Docker Dashboard.

    ![A screenshot of Docker Dashboard showing the running instances of Redis containers](images/running-redis-containers.webp?border=true)

### Run Redis container in a controlled network

By default, containers automatically connect to a special network called a bridge network when you run them. This bridge network acts like a virtual bridge, allowing containers on the same host to communicate with each other while keeping them isolated from the outside world and other hosts. It's a convenient starting point for most container interactions. However, for specific scenarios, you might want more control over the network configuration.

Here's where the `--network` flag with the `docker run` command comes in.

Follow the steps to see how to connect a Redis container to a custom network.

1. Create a network by using the following command:

    ```console
    $ docker network create mynetwork
    ```

2. Verify the network by running the following command:

    ```console
    $ docker network ls
    ```

3. Connect Redis to the existing network by using the following command:

    ```console

>     $ docker run -d -p 6381:6379 –network mynetwork redis

Reviewer: ```suggestion
    $ docker run -d -p 6381:6379 --network mynetwork redis
```

    ```

    This will start Redis container in the background, mapped to the host port 6381 and attached to the `mynetwork` network. You passed the `--network` parameter to override the container default by connecting the container to custom Docker network for better isolation and communication with other containers.

    > **Inspecting the container network**
    >
    > You can use `docker network inspect` command to see if the container is tied to this new bridge network.
    { .tip }

### Manage the resources

By default, containers are not limited in their resource usage. However, on shared systems, it's crucial to manage resources effectively. It's important not to let a running container consume too much of the host machine's memory.

This is where the `docker run` command shines again. It offers flags like `--memory` and `--cpus` to restrict how much CPU and memory a container can use.

```console

> $ docker run -d --memory=”512m” --cpus=”.5” redis

Reviewer: ```suggestion
$ docker run -d --memory="512m" --cpus=".5" redis
```

Sometimes, you might need to override the default commands (`CMD`) or entry points (`ENTRYPOINT`) defined in a Docker image, especially when using Docker Compose.

1. Create a `compose.yaml` file with the following content:

    ```yaml
    services:
        redis:
         image: redis
         entrypoint: ["docker-entrypoint.sh"]
         command: ["redis-server", "--requirepass", "redispassword"]
    ```

    The Compose file defines a service named `redis` that uses the official Redis image, sets an entrypoint script, and starts the container with password authentication.

2. Bring up the service by running the following command:

    ```console
    $ docker compose up -d
    ```

    This command starts the Redis service defined in the Docker Compose file.

3. Verify the authentication with Docker Dashboard.

    Open the Docker Dashboard, select the **Redis** container and select **Exec** to enter into the container shell.

    ![A screenshot of the Docker Dashboard selecting the Redis container and entering into its shell using EXEC button](images/exec-into-redis-container.webp?border=true) 

## Additional resources

> * [Build variables](/build/building/variables/)

Reviewer 2: This page doesn't have anything related to build variables so I don't think we need this link. [This page](https://docs.docker.com/engine/reference/run/) has a lot of overlap with this guide so maybe somewhere there is where you want to link to.

Author: I have removed the link.

The `docker run` command offers a powerful way to override these defaults and tailor the container's behavior to your liking. The command offers several flags that let you to customize container behavior on the fly.

Here's a few ways you can achieve this.

### Overriding the network ports

Sometimes you might want to use separate database instances for development and testing purposes. Running these database instances on the same port might conflict. You can use the `-p` option in `docker run` to map container ports to host ports, allowing you to run the multiple instances of the container without any conflict.

```console
$ docker run -d -p HOST_PORT:CONTAINER_PORT postgres
```

### Setting environment variables

This option sets an environment variable `foo` inside the container with the value `bar`.

```console

> $ docker run -e foo=bar postgres env

Author: If you look at the initial part, it says "For example, if you have an existing database container that listens on the standard port and you want to run a new instance of the same database container, then you might want to change the port settings the new container listens on so that it doesn’t conflict with the existing container." As we're talking about a database container, it does make sense to pick up a database container rather than Alpine. Changing to alpine would disturb the flow of the explanation. Let me know your thoughts.

```

You will see output like the following:

```console
HOSTNAME=2042f2e6ebe4
foo=bar
```

> **Tip**
>
> The `.env` file acts as a convenient way to set environment variables for your Docker containers without cluttering your command line with numerous `-e` flags. To use a `.env` file, you can pass `--env-file` option with the `docker run` command.

> { .tip }

Author: Added to the tip.

1. [Download and install](/get-docker/) Docker Desktop.

### Run multiple instance of the Postgres database

1.  Start a container using the [Postgres image](https://hub.docker.com/_/postgres) with the following command:
    
    ```console
    $ docker run -d -e POSTGRES_PASSWORD=secret -p 5432:5432 postgres
    ```

    This will start the Postgres database in the background, listening on the standard container port `5432` and mapped to port `5432` on the host machine.

2. Start a new Postgres container mapped to a different port. 

    ```console
    $ docker run -d -e POSTGRES_PASSWORD=secret -p 5433:5432 postgres
    ```

    This will start a new Postgres container in the background, listening on the standard container port `5432` but mapped to port `5433` on the host machine. You override the host port just to ensure that this new container doesn't conflict with the existing running container.

3. Verify if both containers are running via the Docker Dashboard.

    ![A screenshot of Docker Dashboard showing the running instances of Postgres containers](images/running-postgres-containers.webp?border=true)

> ### Run Postgres container in a controlled network

Reviewer 2: This section doesn't address the difference between the default bridge and custom bridge networks and DNS resolution. The paragraph seems to suggest that the default bridge is convenient for container intercommunication, but that's only via container ip which isn't very usable. I think this section should demonstrate DNS resolution over a custom bridge instead (not only how to connect to one)

Author: Added a tip/section that discusses customer Vs default bridge network.

Reviewer 2: Thanks!

    > **Inspecting the container network**
    >
    > You can use `docker network inspect` command to see if the container is tied to this new bridge network.
    { .tip }

### Manage the resources

By default, containers are not limited in their resource usage. However, on shared systems, it's crucial to manage resources effectively. It's important not to let a running container consume too much of the host machine's memory.

This is where the `docker run` command shines again. It offers flags like `--memory` and `--cpus` to restrict how much CPU and memory a container can use.

```console
$ docker run -d -e POSTGRES_PASSWORD=secret --memory="512m" --cpus=".5" postgres
```

The `--cpus` flag specifies the CPU quota for the container. Here, it's set to half a CPU core (0.5) whereas the `--memory` flag specifies the memory limit for the container. In this case, it's set to 512 MB.

> ### Override the default `CMD` and `ENTRYPOINT` in Docker Compose

Reviewer 2: I think this section should not only show the compose example, but also overriding default commands for `docker run`.

Author: I added a section for overriding the container default with`docker run`. Do check and let me know if it looks good.

Sometimes, you might need to override the default commands (`CMD`) or entry points (`ENTRYPOINT`) defined in a Docker image, especially when using Docker Compose.

1. Create a `compose.yml` file with the following content:

    ```yaml
    services:
      postgres:
        image: postgres
        entrypoint: ["docker-entrypoint.sh", "postgres"]
        command: ["-h", "localhost", "-p", "5432"]
        environment:
          POSTGRES_PASSWORD: secret 
    ```

    The Compose file defines a service named `postgres` that uses the official Postgres image, sets an entrypoint script, and starts the container with password authentication.

2. Bring up the service by running the following command:

    ```console
    $ docker compose up -d
    ```

    This command starts the Postgres service defined in the Docker Compose file.

3. Verify the authentication with Docker Dashboard.

    Open the Docker Dashboard, select the **Postgres** container and select **Exec** to enter into the container shell. You can type the following command to connect to the Postgres database:

>     ```
>     psql -U postgres

Reviewer 2: ```suggestion
    ```console
    # psql -U postgres
```
